stem/model.py

```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
+                                                                element_type="LINE_2N",
+                                                                node_ids=[start_node_id, end_node_id])
+
+                    # increment the element id
+                    new_element_id += 1
+
+                # add the new mesh to the mesh data
+                self.gmsh_io.mesh_data["physical_groups"][mp.name]["node_ids"] = sorted(list(new_mesh.nodes.keys()))
+                self.gmsh_io.mesh_data["physical_groups"][mp.name]["element_ids"] = \
+                    sorted(list(new_mesh.elements.keys()))
+
+                for element_id, element in new_mesh.elements.items():
+                    self.gmsh_io.mesh_data["elements"]["LINE_2N"][element_id] = element.node_ids
+
+                mp.mesh = new_mesh
+
+    def __get_maximum_element_id(self) -> int:
+        """
+        Returns the maximum element id within the mesh from the mesh data
+
+        Returns:
+            - int: the maximum element id
+
+        """
+        max_element_id = 0
+        for element_type, mesh_element_info in self.gmsh_io.mesh_data["elements"].items():
+            max_element_id = max(max_element_id, max(mesh_element_info.keys()))
+
+        return int(max_element_id)
+
+    def __get_line_string_end_nodes(self, model_part: ModelPart) -> List[List[int]]:
+        """
+        Finds the nodes at both of the line string ends. A line string end is defined as the node which coincides with
+        a geometry point.
+
+        Args:
+            - model_part (:class:`stem.model_part.ModelPart`): model part from which the spring elements need to be
+                extracted.
+
+        Returns:
+            - List[List[int]]: a list of lists which contains both end nodes for separate each line string.
+
+        """
+
+        # assert mesh and geometry are initialised
+        if model_part.geometry is None:
+            raise ValueError(f"Geometry of model part `{model_part.name}` not yet initialised.")
+
+        if model_part.mesh is None:
+            raise ValueError(f"Mesh of model part `{model_part.name}` not yet initialised.")
+
+        # find end nodes
+        end_nodes = self.__find_end_nodes_of_line_strings(model_part.mesh)
+        # find the node ids corresponding to the geometry points
+        node_ids_at_geometry_points = Utils.find_node_ids_close_to_geometry_nodes(
+            mesh=model_part.mesh, geometry=model_part.geometry, eps=1e-06
+        )
+
+        element_ids_search_space = list(model_part.mesh.elements.keys())
+        node_ids_search_space = list(model_part.mesh.nodes.keys())
+
+        # retrieve the connectivity
+        # node -> elements
+        node_to_elements = self.__map_node_to_elements(model_part.mesh)
+
+        # initialise output list
+        line_node_ids = []
+        # initialise a list for end-point we have already encountered n the clustering algorithm
```

Comment:
`        # initialise a list for end-point we have already encountered in the clustering algorithm`